### PR TITLE
Table operator layer

### PR DESCRIPTION
**Pull request by QingSunVT**

### CODE REVIEW

**Comment by luotao1** [comment]
请在最新的develop分支上开发新功能，另外描述下这个Pull Request是做什么的

**Comment by QingSunVT** [comment]
这个layer是读取input matrix的rows。功能类似tableprojection，但读取的是input 而不是parameter

File: paddle/gserver/layers/TableOperatorLayer.cpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+     IVectorPtr ids = getInputLabel(*inputLayers_[0]);
+     MatrixPtr outG = getOutputGrad();
+     MatrixPtr inG = getInputGrad(1);
+     int batch_size = getInput(1).getBatchSize();
+     int r = getSize();
+     int h = inG->getWidth()/r;
+     //
+     inG->reshape(batch_size*h, r);
+     for (int i = 0; i < batch_size; i++) {
+         ids->setElement(i, ids->getElement(i) + i*r);
```
Comment (emailweixu) [inline paddle/gserver/layers/TableOperatorLayer.cpp]: do not recommend to use a loop for setElement or getElement. This is very inefficient for GPU. For GPU, you can create ICpuVector and copy data from gpu vector. And modify over cpu vector using Vector::getData().

File: python/paddle/trainer_config_helpers/layers.py [hunk 2]
```diff
@@ -54,7 +54,7 @@
            'multi_binary_label_cross_entropy',
            'rank_cost', 'lambda_cost', 'huber_cost',
            # 'block_expand_layer',  # TODO(yuyang18): this layer is not correct
-           'out_prod_layer', 'print_layer'
+           'out_prod_layer', 'print_layer','table_operator_layer'
```
Comment (emailweixu) [inline python/paddle/trainer_config_helpers/layers.py]: table_operator_layer => table_operator

File: python/paddle/trainer_config_helpers/layers.py [hunk 3]
```diff
@@ -1430,6 +1431,18 @@ def cos_sim(a, b, scale=5, size=1, name=None, layer_attr=None):
         )
     return LayerOutput(name, LayerType.COSINE_SIM, parents=[a, b])
 
+@wrap_name_default()
+@layer_support()
+def table_operator_layer(a,b,size,name=None,layer_attr=None):
```
Comment (emailweixu) [inline python/paddle/trainer_config_helpers/layers.py]: make it an operator like conv_operator

Comment (emailweixu) [inline python/paddle/trainer_config_helpers/layers.py]: comment the function following examples in this file

File: paddle/gserver/layers/TableOperatorLayer.h [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+#include "Layer.h"
+#include "paddle/math/Matrix.h"
+#include "paddle/utils/ThreadLocal.h"
+
+namespace paddle {
+/**
+ * Select rows from input matrix
+ */
+class TableOperatorLayer : public Layer {
```
Comment (emailweixu) [inline paddle/gserver/layers/TableOperatorLayer.h]: make it an operator like ConvOperator or DotmulOperator.

**Review by emailweixu — COMMENTED** [review]
Also need to change test_LayerGrad to add unittest.

**Comment by luotao1** [comment]
感谢您给PaddlePaddle贡献代码。由于Paddle V1/V2版本已不再维护，相关代码也已从develop分支上删除，因此关闭您的PR，欢迎您向Paddle最新版-Fluid贡献代码。
Thanks for contributing to PaddlePaddle! Since V1/V2 will not be maintained anymore, and related codes have been deleted from develop branch as well, we close this PR. Welcome to contribute to Fluid——the latest version of PaddlePaddle.

**Pull request closed without merging**